module/os/linux/spl/spl-taskq.c

```diff
@@ -1255,12 +1275,92 @@ module_param_call(spl_taskq_kick, param_set_taskq_kick, param_get_uint,
 MODULE_PARM_DESC(spl_taskq_kick,
 	"Write nonzero to kick stuck taskqs to spawn more threads");
 
+static int
+spl_taskq_expand(unsigned int cpu, struct hlist_node *node)
+{
+	taskq_t *tq = list_entry(node, taskq_t, tq_hp_cb_node);
+	unsigned long flags;
+	int err = 0;
+
+	ASSERT(tq);
+	spin_lock_irqsave_nested(&tq->tq_lock, flags, tq->tq_lock_class);
+
+	if (!(tq->tq_flags & TASKQ_ACTIVE))
+		goto out;
+
+	if (tq->tq_flags & TASKQ_THREADS_CPU_PCT) {
+		int nthreads = MIN(tq->tq_orig_maxthreads, 100);
+		nthreads = MAX(nthreads, 0);
+		nthreads = MAX(((num_online_cpus() + 1) * nthreads) / 100, 1);
+		tq->tq_maxthreads = nthreads;
+	} else {
+		tq->tq_maxthreads++;
```

Comment:
Why is the right thing to always add another thread in the `!TASKQ_THREADS_CPU_PCT` case?  Could you explain you're reasoning here, it seem to me like in this case we should simply leave things as they were.

Reply:
We only call this callback on taskqs that are either in the TASKQ_THREADS_CPU_PCT case, or ones where we specifically create the same number of threads as we have boot_ncpus. I used this as shorthand for "taskqs that we want to have the same number of threads as we have cores"; in that case, adding a new core means we should add a new maxthread. I don't know if that logic is ever used, however; we might only use CPU_PCT with 100% for that case. If that's so, we can simply remove this else statement.

Reply:
Ahh, I see.  Then we're guaranteed this callback will be run exactly once for each core which comes online.  That wasn't clear to me until I look a more careful look at the kernel documentation.  Well, in that case I'm fine with keeping it, though a little additional explanation about the expected cases wouldn't hurt.